### PR TITLE
gh-82367: Use `FindFirstFile` Win32 API in `ntpath.realpath()`

### PR DESCRIPTION
**Pull request by moonsikpark**
`ntpath.realpath()` breaks out of traversing a series of paths where a (handled) ERROR_ACCESS_DENIED or ERROR_SHARING_VIOLATION occurs.

Using `FindFirstFile` Win32 API, we can query their real name to correct the case and also resolve MS-DOS (also called 8.3) style file names such as `C:\\LONGFI~1.TXT`.

* Issue: gh-82367

----
:books: Documentation preview :books:: https://cpython-previews--110298.org.readthedocs.build/

### CODE REVIEW

File: Doc/whatsnew/3.13.rst
Comment (zware) [inline Doc/whatsnew/3.13.rst]: This is generally the issue number rather than the PR number, so `82367` rather than `110298`.

File: Misc/NEWS.d/next/Windows/2023-10-03-12-30-59.gh-issue-82367.nxwfMx.rst
Comment (zware) [inline Misc/NEWS.d/next/Windows/2023-10-03-12-30-59.gh-issue-82367.nxwfMx.rst]: The first line here isn't necessary.

File: Doc/library/os.path.rst
Comment (zware) [inline Doc/library/os.path.rst]: I'm not sure this belongs within the 'note' block here.  It might go better in the first paragraph?

It might be good to include an example of what `C:\\LONGFI~1.TXT` could expand into (ideally using a filename that actually exists in a standard installation, but that may be asking too much :)).

Also, it's generally "On Windows" rather than "In Windows".

**Review by zware — COMMENTED** [review]
A few documentation nits, but mostly this looks pretty good to me.  It will need approval from either @zooba or @eryksun before I'd be comfortable merging it, though.

**Comment by zooba** [comment]
The code change looks great to me. Awesome job!

With Zach's requested doc changes, I think this is good to merge.

File: Misc/NEWS.d/next/Windows/2023-10-03-12-30-59.gh-issue-82367.nxwfMx.rst
Comment (moonsikpark) [reply Misc/NEWS.d/next/Windows/2023-10-03-12-30-59.gh-issue-82367.nxwfMx.rst]: removed the line.

File: Doc/whatsnew/3.13.rst
Comment (moonsikpark) [reply Doc/whatsnew/3.13.rst]: changed it to the issue number.

File: Doc/library/os.path.rst
Comment (moonsikpark) [reply Doc/library/os.path.rst]: Moved to the first paragraph and applied suggestion. However... I was unable to find a suitable filename that is short enough and actually exists in a standard installation, so I used  ``C:\\PROGRA~1\\LONGFI~1.TXT`` and ``C:\\Program Files\\LongFile.txt``.

Comment (moonsikpark) [reply Doc/library/os.path.rst]: Oops, `LongFile.txt` is equal to 8.3 so it won't be shortened! Changed to `LongFileName.txt`.

Comment (moonsikpark) [reply Doc/library/os.path.rst]: Actually, `LongFile.txt` is equal to 8.3 so it won't be shortened. Also, writing both C:\\PROGRA~1\\LONGFI~1.TXT and C:\\Program Files\\LongFile.txt seems to be needlessly long :(

Comment (moonsikpark) [reply Doc/library/os.path.rst]: Changed the example with``C:\\PROGRA~1`` to ``C:\\Program Files``.

File: Lib/ntpath.py
Comment (eryksun) [inline Lib/ntpath.py]: The following cases should also be included for `FindFirstFileW()` name normalization.

* The filesystem doesn't support WinAPI `GetFinalPathNameByHandleW()`.
    * 1: `ERROR_INVALID_FUNCTION`
    * 50: `ERROR_NOT_SUPPORTED`
    * 87: `ERROR_INVALID_PARAMETER`
* The type of reparse point is not supported (e.g. appexec links).
    * 1920: `ERROR_CANT_ACCESS_FILE`
* A reparse point can't be resolved, and `_readlink_deep()` fails.
    * 1921: `ERROR_CANT_RESOLVE_FILENAME`

Note that `_readlink_deep()` does nothing to ensure that the result is a final, normalized pathname. It just tries to keep reading symlinks until the target isn't a symlink or access/reading fails in a manner that's allowed. If it succeeds, the resulting pathname may include any number of symlinks, mount points, and short names. `_getfinalpathname_nonstrict()` should instead pass both the pathname and the `seen` set into `_readlink_deep()`. If it succeeds, then add the resulting pathname to the set, and continue the loop to try to get a final, normalized pathname.

Comment (moonsikpark) [reply Lib/ntpath.py]: Will it be fine if I add  `1, 50, 87, 1920, 1921` to the error tuple in line 691?

Also, do you think the fix to address your note about `_readlink_deep()` should also be included in this pr?

Comment (eryksun) [reply Lib/ntpath.py]: Do you want to work on a new issue to ensure that the deep readlink result gets resolved to a final, normalized pathname?

Comment (moonsikpark) [reply Lib/ntpath.py]: Yes, I won't mind it.

I've added the errors to the error tuple.

File: Lib/test/test_ntpath.py
Comment (eryksun) [inline Lib/test/test_ntpath.py:656]: Automatic generation of short names may be disabled on NTFS volumes for the sake of performance. (They're not supported at all on ReFS and exFAT.) Try to set the short name using the following command line:

```python
subprocess.run(['fsutil.exe', 'file', 'setShortName', test_file, 'LONGFI~1.TXT'], creationflags=subprocess.DETACHED_PROCESS)
```

File: Lib/test/test_ntpath.py
Comment (eryksun) [inline Lib/test/test_ntpath.py]: These needs to be a real path for the comparison to be reliable.

```suggestion
        ABSTFN = ntpath.realpath(os_helper.TESTFN)
```

File: Lib/test/test_ntpath.py
Comment (moonsikpark) [reply Lib/test/test_ntpath.py:656]: You're right! Do you think I should check the drive filesystem also?

File: Lib/test/test_ntpath.py
Comment (moonsikpark) [reply Lib/test/test_ntpath.py]: Right. Fixed.

File: Lib/test/test_ntpath.py
Comment (eryksun) [reply Lib/test/test_ntpath.py:656]: Try to run the above command to set the short name, without worry about whether it succeeds. Then if `ntpath.realpath(test_file_short)` isn't the same as `test_file`, skip the test with the message that the filesystem lacks the required support for short filenames.

Comment (moonsikpark) [reply Lib/test/test_ntpath.py:656]: Ah, thus `run()` instead of `check_call()`. OK, wilco.

File: Lib/test/test_ntpath.py
Comment (eryksun) [inline Lib/test/test_ntpath.py]: This should be executed right after creating the file. Use `subprocess.run()` without checking the result.

Comment (moonsikpark) [reply Lib/test/test_ntpath.py]: OK, fixed.

File: Lib/test/test_ntpath.py
Comment (eryksun) [inline Lib/test/test_ntpath.py]: You can't consume the assertion from the second `assertPathEqual()` call. That's the actual test. Everything else is setup and checking whether the test is supported.

```suggestion
        try:
            self.assertPathEqual(test_file, ntpath.realpath(test_file_short))
        except AssertionError:
            raise unittest.SkipTest('the filesystem seems to lack support for short filenames')

        # Deny the right to [S]YNCHRONIZE on the file to all users (SDDL
        # alias "BU") to make opening it fail with ERROR_ACCESS_DENIED. 
        p = subprocess.run(['icacls.exe', test_file, '/deny', '*BU:(S)'],
                creationflags=subprocess.DETACHED_PROCESS)
        if p.returncode:
           raise unittest.SkipTest('failed to deny access to the test file') 

       self.assertPathEqual(test_file, ntpath.realpath(test_file_short))
```

Comment (moonsikpark) [reply Lib/test/test_ntpath.py]: Right. Fixed.

File: Lib/test/test_ntpath.py
Comment (eryksun) [inline Lib/test/test_ntpath.py]: This flag was added as `subprocess.DETACHED_PROCESS` way back in 3.7. There's no need to hard code it anymore.

Comment (moonsikpark) [reply Lib/test/test_ntpath.py]: Right. Removed.

File: Lib/test/test_ntpath.py
Comment (eryksun) [inline Lib/test/test_ntpath.py]: Change the name used here to `_getfinalpathname`. The API call is `GetFinalPathNameByHandleW()`, but what we're actually trying to make fail is opening the handle via `CreateFileW()` in the implementation of `nt._getfinalpathname()`.

Comment (moonsikpark) [reply Lib/test/test_ntpath.py]: Right. Fixed.

**Review by eryksun — APPROVED** [review]
LGTM. Thank you, Moonsik, and thanks also for taking the extra time to make the test more robust.

**Comment by zware** [comment]
Thanks Steve and Eryk for the reviews, and Moonsik for the patch and excellent collaboration!

**Pull request merged**
